dali/kernels/reduce/mean_stddev_gpu_impl.cuh

```diff
@@ -0,0 +1,386 @@
+
+namespace dali {
+namespace kernels {
+namespace reduce_impl {
+
+template <typename Out, typename Scale>
+struct ScaleAndConvert {
+  Scale scale = 1;
+  template <typename T>
+  DALI_HOST_DEV Out operator()(T x) const {
+    return ConvertSat<Out>(x * scale);
+  }
+};
+
+template <typename Out, typename In, typename Actual>
+class MeanImplBase {
+ public:
+  Actual &This() { return static_cast<Actual&>(*this); }
+  const Actual &This() const { return static_cast<const Actual&>(*this); }
+
+  using scale_t = std::conditional_t<std::is_same<Out, double>::value, double, float>;
+  using Postprocessor = ScaleAndConvert<Out, scale_t>;
+
+  Postprocessor *GetPostprocessorsImpl(WorkArea &wa) const {
+    assert(!This().ReduceBatch());
+    int n = This().SimplifiedOutputShape().num_samples();
+    Postprocessor *pp = wa.ParamBuffer<Postprocessor>(n);
+    for (int i = 0; i < n; i++) {
+      DALI_ENFORCE(This().ReducedElements(i) > 0, "Cannot calculate a mean from 0 elements");
+      pp[i].scale = 1.0 / This().ReducedElements(i);
+    }
+    return pp;
+  }
+
+  Postprocessor GetPostprocessorImpl() const {
+    assert(This().ReduceBatch());
+    DALI_ENFORCE(This().TotalReducedElements() > 0, "Cannot calculate a mean from 0 elements");
+    return { static_cast<float>(1.0 / This().TotalReducedElements()) };
+  }
+};
+
+template <typename Out, typename In, typename Acc = default_sum_acc_t<Out, In>>
+class MeanImplGPU : public ReduceImplGPU<Out, In, Acc, MeanImplGPU<Out, In, Acc>>,
+                    public MeanImplBase<Out, In, MeanImplGPU<Out, In, Acc>> {
+ public:
+  using MeanBase = MeanImplBase<Out, In, MeanImplGPU<Out, In, Acc>>;
+  using typename MeanBase::Postprocessor;
+  reductions::sum GetReduction() const { return {}; }
+};
+
+/**
+ * @brief Subtracts a mean value stored in specified memory location and squares the difference
+ */
+template <class Mean>
+struct VarianceIndirect {
+  const Mean *__restrict__ mean = nullptr;
+  template <typename T>
+  DALI_HOST_DEV DALI_FORCEINLINE
+  auto operator()(const T &x) const noexcept {
+    #ifdef __CUDA_ARCH__
+      auto d = x - __ldg(mean);
+    #else
+      auto d = x - *mean;
+    #endif
+      return d * d;
+  }
+};
+
+/**
+ * @brief Returns a `reduce::variance` functor with mean value taken from a tensor
+ */
+template <int non_reduced_ndim, typename Mean>
+struct VariancePreprocessor;
+
+template <typename Mean>
+struct VariancePreprocessor<1, Mean> {
+  const Mean *__restrict__ mean;
+  i64vec<1> stride;
+
+  DALI_HOST_DEV DALI_FORCEINLINE
+  reductions::variance<Mean> Get(const i64vec<1> &pos) const {
+    auto offset = dot(pos, stride);
+  #ifdef __CUDA_ARCH__
+    Mean m = __ldg(mean + offset);
+  #else
+    Mean m = mean[offset];
+  #endif
+    return { m };
+  }
+};
+
+template <typename Mean>
+struct VariancePreprocessor<2, Mean> {
+  const Mean *mean;
+  i64vec<2> stride;
+  /// Calculates the fully reduced inner offset based on non-reduced `pos[1]`
+  DropDims inner_dims;
```

Comment:
This will be nitpicky, but according to the Google Style Guide's nomenclature that we use, I would say this is going into the `class` area, with the members that must be set to make the `Get()` work (especially with inner_dims). So maybe give this a constructor?

Also, is this really a VariancePreprocessor or more of VariancePreprocessorBank?

Reply:
It's a bank. I'll see how many more line breaks it will cause :\

Reply:
Done.